contracts/oracle/CurveOracle.sol

```diff
@@ -13,26 +14,48 @@ contract CurveOracle is IBaseOracle {
   using SafeMath for uint;
 
   IBaseOracle public tokenOracle;
+  ICurveRegistry public registry;
 
-  constructor(IBaseOracle _tokenOracle) public {
+  mapping(address => address[]) public ulTokens; // lpToken -> underlying token array
+  mapping(address => address) public poolOf; // lpToken -> pool
+
+  constructor(IBaseOracle _tokenOracle, ICurveRegistry _registry) public {
     tokenOracle = _tokenOracle;
+    registry = _registry;
+  }
+
+  /// @dev Return pool address given LP token and update pool info if not exist.
+  /// @param lp LP token to find the corresponding pool.
+  function getPool(address lp) public returns (address) {
+    address pool = poolOf[lp];
+    if (pool == address(0)) {
+      require(lp != address(0), 'no lp token');
+      pool = registry.get_pool_from_lp_token(lp);
+      poolOf[lp] = pool;
+      uint n = registry.get_n_coins(pool);
+      address[8] memory tokens = registry.get_coins(pool);
+      ulTokens[lp] = new address[](n);
+      for (uint i = 0; i < n; i++) {
+        ulTokens[lp][i] = tokens[i];
+      }
+    }
+    return pool;
   }
 
   /// @dev Return the value of the given input as ETH per unit, multiplied by 2**112.
-  /// @param token The ERC-20 token to check the value.
-  function getETHPx(address token) external view override returns (uint) {
-    ICurvePool pool = ICurvePool(token);
+  /// @param lp The ERC-20 LP token to check the value.
+  function getETHPx(address lp) external view override returns (uint) {
+    ICurvePool pool = ICurvePool(poolOf[lp]);
     uint minPx = uint(-1);
-    for (uint idx = 0; ; idx++) {
-      try pool.coins(idx) returns (address coin) {
-        uint decimals = IERC20Decimal(coin).decimals();
-        uint tokenPx = tokenOracle.getETHPx(coin);
-        if (decimals < 18) tokenPx = tokenPx.div(10**(18 - decimals));
-        if (decimals > 18) tokenPx = tokenPx.mul(10**(decimals - 18));
-        if (tokenPx < minPx) minPx = tokenPx;
-      } catch (bytes memory) {
-        break;
-      }
+    address[] memory tokens = ulTokens[lp];
+    uint n = tokens.length;
+    for (uint idx = 0; idx < n; idx++) {
+      address token = tokens[idx];
+      uint decimals = IERC20Decimal(token).decimals();
```

Comment:
If we are doing the catching, I think we can do another step and also catch decimals.
```
struct UnderlyingToken {
   uint8 decimals;
   address token;
}
```
^ this fits into a single word. It won't work with token that decimals can change, but I don't think those exist.